File: ios/AnyAICLIRemoteCore/InteractionAnswerCodec.swift
import Foundation

public enum InteractionAnswerCodec {
  public static func result(_ answer: InteractionAnswer) -> [String: Any] {
    switch answer {
    case .accept(let answers, let annotations):
      var payload: [String: Any] = ["outcome": "accepted", "answers": answers]
      let encodedAnnotations = annotations.reduce(into: [String: [String: String]]()) { result, entry in
        let notes = entry.value.notes.trimmingCharacters(in: .whitespacesAndNewlines)
        if !notes.isEmpty { result[entry.key] = ["notes": notes] }
      }
      if !encodedAnnotations.isEmpty { payload["annotations"] = encodedAnnotations }
      return payload
    case .chatAbout(let values):
      return ["outcome": "chat_about_this", "partialAnswers": values]
    case .skipInterview(let values):
      return ["outcome": "skip_interview", "partialAnswers": values]
    case .approve:
      return ["outcome": "approved"]
    case .cancel(let feedback):
      var payload: [String: Any] = ["outcome": "cancelled"]
      let trimmedFeedback = feedback.trimmingCharacters(in: .whitespacesAndNewlines)
      if !trimmedFeedback.isEmpty {
        payload["feedback"] = trimmedFeedback
      }
      return payload
    case .cancelAsk:
      return ["outcome": "cancelled"]
    case .abandon:
      return ["outcome": "abandoned"]
    }
  }
}

File: android/core/session/src/main/java/com/anyaicliremote/core/session/InteractionAnswerCodec.kt
package com.anyaicliremote.core.session

import com.anyaicliremote.core.model.InteractionAnnotation
import com.anyaicliremote.core.model.InteractionAnswer
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * Encodes a typed [InteractionAnswer] into the daemon's neutral interaction
 * response. The daemon denormalizes it into the provider shape; the client keeps
 * no provider wire knowledge here.
 */
object InteractionAnswerCodec {
    fun toResult(answer: InteractionAnswer): JsonObject = when (answer) {
        is InteractionAnswer.Accept -> buildJsonObject {
            put("outcome", "accepted")
            put("answers", answersObject(answer.answers))
            val annotations = annotationsObject(answer.annotations)
            if (annotations.isNotEmpty()) put("annotations", annotations)
        }
        is InteractionAnswer.ChatAbout -> partial("chat_about_this", answer.partialAnswers)
        is InteractionAnswer.SkipInterview -> partial("skip_interview", answer.partialAnswers)
        InteractionAnswer.CancelAsk -> buildJsonObject { put("outcome", "cancelled") }
        InteractionAnswer.Approve -> buildJsonObject { put("outcome", "approved") }
        is InteractionAnswer.Cancel -> buildJsonObject {
            put("outcome", "cancelled")
            if (answer.feedback.isNotBlank()) put("feedback", answer.feedback)
        }
        InteractionAnswer.Abandon -> buildJsonObject { put("outcome", "abandoned") }
    }

    private fun answersObject(answers: Map<String, List<String>>): JsonObject = buildJsonObject {
        answers.forEach { (index, labels) ->
            put(index, JsonArray(labels.map(::JsonPrimitive)))
        }
    }

    // Per-question notes, keyed by the same decimal question index as answers.
    private fun annotationsObject(annotations: Map<String, InteractionAnnotation>): JsonObject = buildJsonObject {
        annotations.forEach { (index, annotation) ->
            val notes = annotation.notes.trim()
            if (notes.isNotEmpty()) put(index, buildJsonObject { put("notes", notes) })
        }
    }

    private fun partial(outcome: String, partialAnswers: Map<String, String>): JsonObject = buildJsonObject {
        put("outcome", outcome)
        put("partialAnswers", buildJsonObject { partialAnswers.forEach { (index, text) -> put(index, text) } })
    }
}
